### PR TITLE
[TC-959] Refactor placement engine to use promises

### PR DESCRIPTION
**Pull request by nullobject**
This PR refactors the placement engine to use plain old promises, rather than signals.

I originally chose signals for the architecture, because I thought that the placement engine needed to be reactive. It turns out that most of the reactive functionality ended up being removed, so signals just became cognitive overhead. Now that we are using promises, everything is much simpler.

I also shuffled a few things between the `placementEngine` and `placePromos` functions to make the API a bit clearer.

## How to Test

I have already [created a PR](https://github.com/conversation/tc/pull/8952) to upgrade to the new promos-client in TC. Fire it up and check that the promos all still work as intended.

### CODE REVIEW

File: src/placePromos.js [hunk 1]
```diff
@@ -19,15 +19,17 @@ import { age } from './utils'
...
- * context will be available to the query.
+ * Extends the given context with functions and objects to be made available to
+ * the query.
  *
  * @private
  */
-function createContext (user, window) {
-  const uaParser = new UAParser(get('navigator.userAgent', window))
+function extendContext (context) {
+  const uaParser = new UAParser(get('navigator.userAgent', context.window))
```
Comment (nickbrowne) [inline src/placePromos.js]: I think not shadowing window is a good idea :)

Comment (nullobject) [reply src/placePromos.js]: I thought so too 😄

File: src/placementEngine.js [hunk 2]
```diff
@@ -1,42 +1,135 @@
...
 
-    // Emit the placed promos and callback functions.
-    .map(({ promos }) => ({ promos, onClick, onClose }))
+  // Place the promos.
+  const context = createContext(userAgent, user)
+  const placedPromos = placePromos(context, promos)
 
-  return stateSignal
+  // Return a promise containing the placed promos, and the callback functions.
+  return Promise.resolve({ promos: placedPromos, onClick, onClose, onView })
```
Comment (nickbrowne) [inline src/placementEngine.js]: I suppose we might want to reject if some unrecoverable error happens, but this seems fine for now. It's at least a much nicer API with flexibility for the future.

**Comment by nickbrowne** [comment]
Really nice improvements, tried the branch out on TC and it works a treat :+1:

**Pull request merged**